registry/handlers/layerupload.go

```diff
@@ -11,154 +11,124 @@ import (
 	ctxu "github.com/docker/distribution/context"
 	"github.com/docker/distribution/digest"
 	"github.com/docker/distribution/registry/api/v2"
-	"github.com/gorilla/handlers"
 )
 
 // layerUploadDispatcher constructs and returns the layer upload handler for
 // the given request context.
-func layerUploadDispatcher(ctx *Context, r *http.Request) http.Handler {
-	luh := &layerUploadHandler{
-		Context: ctx,
-		UUID:    getUploadUUID(ctx),
+func layerUploadHandler(ctx *Context, w http.ResponseWriter, r *http.Request) (err error) {
+	switch r.Method {
+	case "POST":
+		err = StartLayerUpload(ctx, w, r)
+	case "GET", "HEAD":
+		err = GetUploadStatus(ctx, w, r)
+	case "PUT":
+		err = PutLayerUploadComplete(ctx, w, r)
+	case "DELETE":
+		err = CancelLayerUpload(ctx, w, r)
+	// TODO(stevvooe): Must implement patch support.
+	// case "PATCH":
+	//	PutLayerChunk(ctx, w, r)
+	default:
+		w.WriteHeader(http.StatusMethodNotAllowed)
 	}
+	return err
+}
 
-	handler := http.Handler(handlers.MethodHandler{
-		"POST": http.HandlerFunc(luh.StartLayerUpload),
-		"GET":  http.HandlerFunc(luh.GetUploadStatus),
-		"HEAD": http.HandlerFunc(luh.GetUploadStatus),
-		// TODO(stevvooe): Must implement patch support.
-		// "PATCH":    http.HandlerFunc(luh.PutLayerChunk),
-		"PUT":    http.HandlerFunc(luh.PutLayerUploadComplete),
-		"DELETE": http.HandlerFunc(luh.CancelLayerUpload),
-	})
-
-	if luh.UUID != "" {
-		state, err := hmacKey(ctx.Config.HTTP.Secret).unpackUploadState(r.FormValue("_state"))
+func layerUploadInfo(ctx *Context, r *http.Request) (uuid string, upload distribution.LayerUpload, state layerUploadState, httpErr error) {
```

Comment:
`layerUploadState` need not be returned here. Just place the returned upload instance into the described state.